Bug: No groups found

This is an intermittent issue where the groups feed disappears after the page has already been loaded. It can only be restored by manually refreshing the page. I don't know how to reproduce it, but it happens often.

![Image](https://github.com/user-attachments/assets/061a49da-b461-4000-b03a-15bd8f6db51d)

Caught it with the console open: 

pomf2.lain.la/f/9dl3hv9h.jpg:1 
       GET https://pomf2.lain.la/f/9dl3hv9h.jpg net::ERR_HTTP2_PROTOCOL_ERROR

pomf2.lain.la/f/9dl3hv9h.jpg:1 
       GET https://pomf2.lain.la/f/9dl3hv9h.jpg net::ERR_HTTP2_PROTOCOL_ERROR

:8080/group/34550%3A932614571afcbad4d17a191ee281e39eebbb41b93fac8fd87829622aeb112f4d%3Aand-other-stuff:1 Access to fetch at 'https://api.allorigins.win/get?url=https%3A%2F%2Fgetstacks.dev%2F' from origin 'http://localhost:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

LinkPreview.tsx:31 
       GET https://api.allorigins.win/get?url=https%3A%2F%2Fgetstacks.dev%2F net::ERR_FAILED 500 (Internal Server Error)

LinkPreview.tsx:73 Error fetching link preview: TypeError: Failed to fetch
    at fetchMetadata (LinkPreview.tsx:31:32)
    at LinkPreview.tsx:81:7

pomf2.lain.la/f/9dl3hv9h.jpg:1 
              GET https://pomf2.lain.la/f/9dl3hv9h.jpg net::ERR_HTTP2_PROTOCOL_ERROR

![Image](https://github.com/user-attachments/assets/f0f5f8ea-7eb9-4172-bdf5-503ac5525a73)

--- a/src/pages/Groups.tsx
+++ b/src/pages/Groups.tsx
@@ -41,14 +41,29 @@ export default function Groups() {
     }
   }, [activeTab]);
 
-  // Fetch all communities
+  // Fetch all communities with improved error handling and timeout
   const { data: allGroups, isLoading: isGroupsLoading } = useQuery({
     queryKey: ["communities"],
     queryFn: async (c) => {
-      const signal = AbortSignal.any([c.signal, AbortSignal.timeout(5000)]);
-      const events = await nostr.query([{ kinds: [34550], limit: 100 }], { signal });
-      return events;
+      try {
+        // Increase timeout to 8 seconds to allow more time for relays to respond
+        const signal = AbortSignal.any([c.signal, AbortSignal.timeout(8000)]);
+        const events = await nostr.query([{ kinds: [34550], limit: 100 }], { signal });
+        
+        // Ensure we always return an array, even if the query fails
+        return Array.isArray(events) ? events : [];
+      } catch (error) {
+        console.error("Error fetching communities:", error);
+        // Return empty array on error instead of throwing
+        return [];
+      }
     },
+    // Add retry logic for better reliability
+    retry: 3,
+    retryDelay: 1000,
+    // Ensure stale data is shown while revalidating
+    staleTime: 60000, // 1 minute
+    gcTime: 300000, // 5 minutes
   });
 
   // Get user's groups
@@ -87,7 +102,7 @@ export default function Groups() {
 
   // Filter and sort all groups
   const sortedAndFilteredGroups = useMemo(() => {
-    if (!allGroups) return [];
+    if (!allGroups || allGroups.length === 0) return [];
 
     // Function to check if a group matches the search query
     const matchesSearch = (community: NostrEvent) => {
@@ -107,43 +122,59 @@ export default function Groups() {
       );
     };
 
-    return allGroups
+    // Create a stable copy of the array to avoid mutation issues
+    const stableGroups = [...allGroups];
+
+    return stableGroups
       .filter(matchesSearch)
       .sort((a, b) => {
-        const aId = getCommunityId(a);
-        const bId = getCommunityId(b);
+        // Ensure both a and b are valid objects
+        if (!a || !b) return 0;
+        
+        try {
+          const aId = getCommunityId(a);
+          const bId = getCommunityId(b);
 
-        const aIsPinned = isGroupPinned(aId);
-        const bIsPinned = isGroupPinned(bId);
+          const aIsPinned = isGroupPinned(aId);
+          const bIsPinned = isGroupPinned(bId);
 
-        // First priority: pinned groups
-        if (aIsPinned && !bIsPinned) return -1;
-        if (!aIsPinned && bIsPinned) return 1;
+          // First priority: pinned groups
+          if (aIsPinned && !bIsPinned) return -1;
+          if (!aIsPinned && bIsPinned) return 1;
 
-        const aIsMember = userMembershipMap.has(aId);
-        const bIsMember = userMembershipMap.has(bId);
+          const aIsMember = userMembershipMap.has(aId);
+          const bIsMember = userMembershipMap.has(bId);
 
-        // Second priority: groups that the user is a member of
-        if (aIsMember && !bIsMember) return -1;
-        if (!aIsMember && bIsMember) return 1;
+          // Second priority: groups that the user is a member of
+          if (aIsMember && !bIsMember) return -1;
+          if (!aIsMember && bIsMember) return 1;
 
-        // If both are pinned or both are not pinned and both are member or both are not member,
-        // sort alphabetically by name
-        const aNameTag = a.tags.find(tag => tag[0] === "name");
-        const bNameTag = b.tags.find(tag => tag[0] === "name");
+          // If both are pinned or both are not pinned and both are member or both are not member,
+          // sort alphabetically by name
+          const aNameTag = a.tags.find(tag => tag[0] === "name");
+          const bNameTag = b.tags.find(tag => tag[0] === "name");
 
-        const aName = aNameTag ? aNameTag[1].toLowerCase() : "";
-        const bName = bNameTag ? bNameTag[1].toLowerCase() : "";
+          const aName = aNameTag ? aNameTag[1].toLowerCase() : "";
+          const bName = bNameTag ? bNameTag[1].toLowerCase() : "";
 
-        return aName.localeCompare(bName);
+          return aName.localeCompare(bName);
+        } catch (error) {
+          console.error("Error sorting groups:", error);
+          return 0;
+        }
       });
   }, [allGroups, searchQuery, isGroupPinned, userMembershipMap]);
 
-  // Loading state skeleton
+  // Loading state skeleton with stable keys
+  const skeletonKeys = useMemo(() => 
+    Array.from({ length: 12 }).map((_, index) => `skeleton-group-${index}`),
+    []
+  );
+  
   const renderSkeletons = () => (
     <div className="grid grid-cols-1 xs:grid-cols-2 sm:grid-cols-2 lg:grid-cols-3 gap-2 sm:gap-3">
-      {Array.from({ length: 12 }).map((_, index) => (
-        <Card key={`skeleton-group-${index}-${Date.now()}`} className="overflow-hidden flex flex-col h-[140px]">
+      {skeletonKeys.map((key) => (
+        <Card key={key} className="overflow-hidden flex flex-col h-[140px]">
           <CardHeader className="flex flex-row items-center space-y-0 gap-3 pt-4 pb-2 px-3">
             <Skeleton className="h-12 w-12 rounded-md" />
             <div className="space-y-1 flex-1">
@@ -209,29 +240,35 @@ export default function Groups() {
           <TabsContent value="posts" className="space-y-4">
             {isGroupsLoading || isUserGroupsLoading ? (
               renderSkeletons()
-            ) : sortedAndFilteredGroups.length > 0 ? (
+            ) : allGroups && sortedAndFilteredGroups && sortedAndFilteredGroups.length > 0 ? (
               <div className="grid grid-cols-1 xs:grid-cols-2 sm:grid-cols-2 lg:grid-cols-3 gap-2 sm:gap-3">
                 {sortedAndFilteredGroups.map((community) => {
-                  const communityId = getCommunityId(community);
-                  const isPinned = isGroupPinned(communityId);
-                  const userRole = userMembershipMap.get(communityId);
-                  const isMember = userMembershipMap.has(communityId);
-                  const stats = communityStats ? communityStats[communityId] : undefined;
+                  if (!community) return null;
+                  try {
+                    const communityId = getCommunityId(community);
+                    const isPinned = isGroupPinned(communityId);
+                    const userRole = userMembershipMap.get(communityId);
+                    const isMember = userMembershipMap.has(communityId);
+                    const stats = communityStats ? communityStats[communityId] : undefined;
 
-                  return (
-                    <GroupCard
-                      key={community.id}
-                      community={community}
-                      isPinned={isPinned}
-                      pinGroup={pinGroup}
-                      unpinGroup={unpinGroup}
-                      isUpdating={isUpdating}
-                      stats={stats}
-                      isLoadingStats={isLoadingStats}
-                      isMember={isMember}
-                      userRole={userRole}
-                    />
-                  );
+                    return (
+                      <GroupCard
+                        key={`${community.id}-${communityId}`}
+                        community={community}
+                        isPinned={isPinned}
+                        pinGroup={pinGroup}
+                        unpinGroup={unpinGroup}
+                        isUpdating={isUpdating}
+                        stats={stats}
+                        isLoadingStats={isLoadingStats}
+                        isMember={isMember}
+                        userRole={userRole}
+                      />
+                    );
+                  } catch (error) {
+                    console.error("Error rendering group card:", error);
+                    return null;
+                  }
                 })}
               </div>
             ) : searchQuery ? (
@@ -259,29 +296,35 @@ export default function Groups() {
           <TabsContent value="my-groups" className="space-y-4">
             {isGroupsLoading || isUserGroupsLoading ? (
               renderSkeletons()
-            ) : userGroups && userGroups.allGroups.length > 0 ? (
+            ) : userGroups && userGroups.allGroups && userGroups.allGroups.length > 0 ? (
               <div className="grid grid-cols-1 xs:grid-cols-2 sm:grid-cols-2 lg:grid-cols-3 gap-2 sm:gap-3">
                 {userGroups.allGroups.map((community) => {
-                  const communityId = getCommunityId(community);
-                  const isPinned = isGroupPinned(communityId);
-                  const userRole = userMembershipMap.get(communityId);
-                  const isMember = userMembershipMap.has(communityId);
-                  const stats = communityStats ? communityStats[communityId] : undefined;
+                  if (!community) return null;
+                  try {
+                    const communityId = getCommunityId(community);
+                    const isPinned = isGroupPinned(communityId);
+                    const userRole = userMembershipMap.get(communityId);
+                    const isMember = userMembershipMap.has(communityId);
+                    const stats = communityStats ? communityStats[communityId] : undefined;
 
-                  return (
-                    <GroupCard
-                      key={community.id}
-                      community={community}
-                      isPinned={isPinned}
-                      pinGroup={pinGroup}
-                      unpinGroup={unpinGroup}
-                      isUpdating={isUpdating}
-                      stats={stats}
-                      isLoadingStats={isLoadingStats}
-                      isMember={isMember}
-                      userRole={userRole}
-                    />
-                  );
+                    return (
+                      <GroupCard
+                        key={`${community.id}-${communityId}`}
+                        community={community}
+                        isPinned={isPinned}
+                        pinGroup={pinGroup}
+                        unpinGroup={unpinGroup}
+                        isUpdating={isUpdating}
+                        stats={stats}
+                        isLoadingStats={isLoadingStats}
+                        isMember={isMember}
+                        userRole={userRole}
+                      />
+                    );
+                  } catch (error) {
+                    console.error("Error rendering group card in my-groups:", error);
+                    return null;
+                  }
                 })}
               </div>
             ) : (
@@ -300,26 +343,32 @@ export default function Groups() {
             ) : allGroups && allGroups.length > 0 ? (
               <div className="grid grid-cols-1 xs:grid-cols-2 sm:grid-cols-2 lg:grid-cols-3 gap-2 sm:gap-3">
                 {allGroups.map((community) => {
-                  const communityId = getCommunityId(community);
-                  const isPinned = isGroupPinned(communityId);
-                  const userRole = userMembershipMap.get(communityId);
-                  const isMember = userMembershipMap.has(communityId);
-                  const stats = communityStats ? communityStats[communityId] : undefined;
+                  if (!community) return null;
+                  try {
+                    const communityId = getCommunityId(community);
+                    const isPinned = isGroupPinned(communityId);
+                    const userRole = userMembershipMap.get(communityId);
+                    const isMember = userMembershipMap.has(communityId);
+                    const stats = communityStats ? communityStats[communityId] : undefined;
 
-                  return (
-                    <GroupCard
-                      key={community.id}
-                      community={community}
-                      isPinned={isPinned}
-                      pinGroup={pinGroup}
-                      unpinGroup={unpinGroup}
-                      isUpdating={isUpdating}
-                      stats={stats}
-                      isLoadingStats={isLoadingStats}
-                      isMember={isMember}
-                      userRole={userRole}
-                    />
-                  );
+                    return (
+                      <GroupCard
+                        key={`${community.id}-${communityId}`}
+                        community={community}
+                        isPinned={isPinned}
+                        pinGroup={pinGroup}
+                        unpinGroup={unpinGroup}
+                        isUpdating={isUpdating}
+                        stats={stats}
+                        isLoadingStats={isLoadingStats}
+                        isMember={isMember}
+                        userRole={userRole}
+                      />
+                    );
+                  } catch (error) {
+                    console.error("Error rendering group card in all-groups:", error);
+                    return null;
+                  }
                 })}
               </div>
             ) : (

